mlir/lib/Dialect/Rock/Transforms/EmulateNarrowType.cpp

```diff
@@ -0,0 +1,256 @@
+    Value newIndex = rewriter.create<arith::DivUIOp>(
+        loc, adaptor.getIndices()[0], scaleConst);
+    // Note: if you're using sgpr offset for some reason, this won't work.
+    Value newValue =
+        rewriter.create<vector::BitCastOp>(loc, newType, adaptor.getValue());
+    rewriter.replaceOpWithNewOp<amdgpu::RawBufferStoreOp>(
+        op, newValue, adaptor.getMemref(), newIndex,
+        adaptor.getBoundsCheckAttr(), nullptr, nullptr);
+    return success();
+  }
+};
+} // end namespace
+
+void RockEmulateNarrowTypePass::runOnOperation() {
+  func::FuncOp op = getOperation();
+  MLIRContext *ctx = &getContext();
+
+  RewritePatternSet prePatterns(ctx);
+  // (Needed because the lowering on rank-0 tensors creates vector.broadcasts,
+  // which LLVM conversion doesn't understand).
+  vector::populateVectorBroadcastLoweringPatterns(prePatterns);
+  // Lower vector.transfer_{read,write} to vector.{load,store} here to let the
+  // vector patterns go through.
+  vector::populateVectorTransferLoweringPatterns(prePatterns,
+                                                 /*maxTransferRank=*/1);
+  if (failed(applyPatternsAndFoldGreedily(op, std::move(prePatterns))))
+    return signalPassFailure();
+
+  arith::NarrowTypeEmulationConverter typeConverter(/*targetBitwidth=*/8);
+  memref::populateMemRefNarrowTypeEmulationConversions(typeConverter);
+  mhal::populateMHalNarrowTypeEmulationConversions(typeConverter);
+
+  auto opLegalCallback = [&typeConverter](Operation *op) {
+    return typeConverter.isLegal(op);
+  };
+  ConversionTarget boundaryTarget(*ctx);
+  boundaryTarget.addDynamicallyLegalOp<func::FuncOp>(
+      [&typeConverter](func::FuncOp op) {
+        return typeConverter.isLegal(op.getFunctionType());
+      });
+  boundaryTarget.addDynamicallyLegalOp<func::CallOp, func::ReturnOp>(
+      opLegalCallback);
+
+  ConversionTarget target(*ctx);
+  target.addDynamicallyLegalDialect<
+      vector::VectorDialect, memref::MemRefDialect, amdgpu::AMDGPUDialect,
+      affine::AffineDialect, arith::ArithDialect>(opLegalCallback);
+  target.addDynamicallyLegalOp<rock::GpuAllocOp>(opLegalCallback);
+
+  // First, we do the conversions on function signatures so that we can get some
+  // unrealized_conversion_cast ops that'll cancel out later.
+  RewritePatternSet boundaryPatterns(ctx);
+  arith::populateArithNarrowTypeEmulationPatterns(typeConverter,
+                                                  boundaryPatterns);
+  mhal::populateMHalNarrowTypeEmulationBoundaryPatterns(typeConverter,
+                                                        boundaryPatterns);
+  if (failed(applyPartialConversion(op, boundaryTarget,
+                                    std::move(boundaryPatterns))))
+    return signalPassFailure();
+
+  RewritePatternSet patterns(ctx);
+  memref::populateMemRefNarrowTypeEmulationPatterns(typeConverter, patterns);
+  vector::populateVectorNarrowTypeEmulationPatterns(typeConverter, patterns);
```

Comment:
Why cant we run the seperate passes ?

Reply:
... because they all share a common type converter and, more fundamentally, there _isn't_ a separate pass. (There's a test pass, but, well, that's for testing)

This is us embracing the developing expectation that downstream projects build their own pass - or at least that makes sense given that all the code came from IREE

Reply:
deferring this to later based on the recent discussion on how upstream could be doing this.